libc/misc/stat.c

```diff
@@ -0,0 +1,1054 @@
+
+	unlink(symPath);
+
+	TEST_ASSERT_EQUAL_INT(0, symlink(path, symPath));
+
+	TEST_ASSERT_EQUAL_INT(0, stat(symPath, &buffer));
+	TEST_ASSERT_TRUE((buffer.st_mode & S_IFMT) == S_IFREG);
+	TEST_ASSERT_EQUAL_INT(permissions_set, buffer.st_mode);
+
+	sym_fd = open(symPath, O_RDONLY);
+	TEST_ASSERT_EQUAL_INT(0, fstat(sym_fd, &buffer));
+	TEST_ASSERT_TRUE((buffer.st_mode & S_IFMT) == S_IFREG);
+	TEST_ASSERT_EQUAL_INT(permissions_set, buffer.st_mode);
+
+	close(sym_fd);
+	unlink(symPath);
+	close(fd);
+	remove(path);
+
+	/* Testing with the lowest possible permissions, have to be 0400 to be accessible by fstat() */
+	fd = open(path, O_CREAT, 0400);
+
+	permissions_set = (NONE_MODE | 0400) & ~mask;
+
+	TEST_ASSERT_EQUAL_INT(0, symlink(path, symPath));
+
+	TEST_ASSERT_EQUAL_INT(0, stat(symPath, &buffer));
+	TEST_ASSERT_TRUE((buffer.st_mode & S_IFMT) == S_IFREG);
+	TEST_ASSERT_EQUAL_INT(permissions_set, buffer.st_mode);
+
+	sym_fd = open(symPath, O_RDONLY);
+	TEST_ASSERT_EQUAL_INT(0, fstat(sym_fd, &buffer));
+	TEST_ASSERT_TRUE((buffer.st_mode & S_IFMT) == S_IFREG);
+	TEST_ASSERT_EQUAL_INT(permissions_set, buffer.st_mode);
+
+	close(sym_fd);
+	unlink(symPath);
+	close(fd);
+	remove(path);
+}
+
+
+TEST(stat_mode, symlink_lstat)
+{
+/* Disabled because of #743 issue: https://github.com/phoenix-rtos/phoenix-rtos-project/issues/743 */
+#ifdef __phoenix__
+	TEST_IGNORE_MESSAGE("#743 issue");
+#endif
+	struct stat buffer;
+	int permissions_set;
+	mode_t mask;
+
+	fd = open(path, O_CREAT, 0666);
+
+	/*
+	 * We subtract the umask because, when we are setting in open() any mask, in reality it is set to 'any_mask & ~umask'
+	 */
+	mask = umask(0);
+	permissions_set = (NONE_MODE | 0666) & ~mask;
+	umask(mask);
+
+	unlink(symPath);
+	TEST_ASSERT_EQUAL_INT(0, symlink(path, symPath));
+
+	TEST_ASSERT_EQUAL_INT(0, lstat(path, &buffer));
+	TEST_ASSERT_TRUE((buffer.st_mode & S_IFMT) == S_IFREG);
+	TEST_ASSERT_EQUAL_INT(permissions_set, buffer.st_mode);
+	TEST_ASSERT_EQUAL_INT(0, buffer.st_size);
+
+	permissions_set = LINK_MODE;
+
+	TEST_ASSERT_EQUAL_INT(0, lstat(symPath, &buffer));
+	TEST_ASSERT_TRUE((buffer.st_mode & S_IFMT) == S_IFLNK);
+	TEST_ASSERT_EQUAL_INT(permissions_set, buffer.st_mode);
+	TEST_ASSERT_EQUAL_INT(strlen(path), buffer.st_size);
+
+	close(fd);
+	remove(path);
+	unlink(symPath);
+
+	/* Testing with the lowest possible permissions */
+
+	fd = open(path, O_CREAT, 0000);
+
+	permissions_set = (NONE_MODE | 0000) & ~mask;
+
+	TEST_ASSERT_EQUAL_INT(0, symlink(path, symPath));
+
+	TEST_ASSERT_EQUAL_INT(0, lstat(path, &buffer));
+	TEST_ASSERT_TRUE((buffer.st_mode & S_IFMT) == S_IFREG);
+	TEST_ASSERT_EQUAL_INT(permissions_set, buffer.st_mode);
+	TEST_ASSERT_EQUAL_INT(0, buffer.st_size);
+
+	permissions_set = LINK_MODE;
+
+	TEST_ASSERT_EQUAL_INT(0, lstat(symPath, &buffer));
+	TEST_ASSERT_TRUE((buffer.st_mode & S_IFMT) == S_IFLNK);
+	TEST_ASSERT_EQUAL_INT(permissions_set, buffer.st_mode);
+	TEST_ASSERT_EQUAL_INT(strlen(path), buffer.st_size);
```

Comment:
As there are test cases intended to check st_mode we can skip checking st_size. Above that we could ignore only asserting the whole st_mode, because of an issue and enable the part with S_IFREG/S_IFLNK now. Applies to the other cases.